src/main/java/bio/overture/ego/security/OAuth2SsoFilter.java

```diff
@@ -61,6 +69,53 @@ public void onAuthenticationSuccess(
   private OrcidService orcidService;
   private GithubService githubService;
 
+  private AuthenticationFailureHandler failureHandler =
+      (request, response, exception) -> {
+        val rootExceptionThrowable = exception.getCause();
+        val application =
+            applicationService.getByClientId(
+                (String) request.getSession().getAttribute("ego_client_id"));
+        String errorRedirectUri =
+            (String) request.getSession().getAttribute("ego_error_redirect_uri");
+
+        val errorRedirect = Redirects.getErrorRedirectUri(application, errorRedirectUri);
+        URIBuilder errorUri;
+
+        if (rootExceptionThrowable instanceof NoPrimaryEmailException) {
+          try {
+            val reqUri = new ArrayList<String>(asList(request.getRequestURI().split("/")));
+            val provider = reqUri.get(reqUri.size() - 1).toUpperCase();
+            errorUri = new URIBuilder(errorRedirect);
+            errorUri.addParameter("error_code", "403");
+            errorUri.addParameter("error_type", rootExceptionThrowable.getMessage());
+            try {
+              ProviderType.resolveProviderType(provider);
+              errorUri.addParameter("provider_type", provider);
+            } catch (IllegalArgumentException e) {
+              log.warn(format("Invalid provider: '%s'", provider));
+            }
+            response.setStatus(403);
+            response.sendRedirect(errorUri.build().toString());
+          } catch (URISyntaxException e) {
+            log.error("Invalid redirect uri");
+          }
+        } else if (rootExceptionThrowable instanceof UserDeniedAuthorizationException) {
+          // User denies Ego access/cancels login
```

Comment:
Thank you for this comment, I thought the exception implied that the user was denied, not that the user did the denying. Maybe the name is not clear enough. `DeniedByUserException` or something?

Reply:
this is a built in exception from Spring :)

Reply:
Alrighty. GOod comment then.